### PR TITLE
Java - Jackson

### PR DESCRIPTION
**Pull request by mcjp78**
Now uses Jackson (Json parser)  for the java driver

### CODE REVIEW

**Comment by PeterKnego** [comment]
Does this break existing API? If so then this should be optional.

File: drivers/java/src/main/java/com/rethinkdb/RethinkDB.java [hunk 1]
```diff
@@ -8,9 +9,16 @@
...
     public Connection.Builder connection() {
         return Connection.build();
     }
+
+    public static void setObjectMapper(ObjectMapper mapper1){
+        mapper = mapper1;
+    }
+    public static ObjectMapper getObjectMapper(){
+        if (mapper != null) return mapper;
+        return new ObjectMapper();
```
Comment (deontologician) [inline drivers/java/src/main/java/com/rethinkdb/RethinkDB.java]: If we're going to create a new mapper, we should probably assign it to `mapper` before returning it.

**Comment by deontologician** [comment]
@PeterKnego I'm a bit torn on this. So in an ideal world, we'd be backwards compatible for a while. In practice though, this is a big maintenance burden and obviously now there's less manpower available for it.

Given that Jackson is better at this stuff than the hand-rolled system we have now, I think it's reasonable to replace it, even if it results in some small changes in client code.

**Comment by mcjp78** [comment]
Right, I agree on POJO tests, I will run some. I can also just set the object mapper to a new instance and then return that.

**Comment by mcjp78** [comment]
Update: ID annotation fixed and tested. POJO's tested. ObjectMapper get and set methods revised.

@deontologician its ready

**Comment by deontologician** [comment]
Ok, this looks pretty good, but I'm going to run the tests before merging, which might take a little time since I haven't got it set up on my laptop here. Bear with me

File: drivers/java/src/main/java/com/rethinkdb/RethinkDB.java [hunk 2]
```diff
@@ -8,9 +9,15 @@
     /**
      * The Singleton to use to begin interacting with RethinkDB Driver
      */
-    public static final RethinkDB r = new RethinkDB();
```
Comment (deontologician) [inline drivers/java/src/main/java/com/rethinkdb/RethinkDB.java]: This should be re-added

**Comment by PeterKnego** [comment]
As noted: this will break stuff for a lot of people. You should not break APIs on the minor version, ideally even major versions should be backwards compatible.

File: drivers/java/src/main/java/com/rethinkdb/net/Util.java [hunk 3]
```diff
@@ -77,94 +78,10 @@ public static String fromUTF8(byte[] ba) {
...
-    }
-
-    private static boolean areParametersMatching(Parameter[] parameters, Map<String, Object> values) {
-        return Arrays.stream(parameters).allMatch(parameter ->
-                values.containsKey(parameter.getName()) &&
-                        values.get(parameter.getName()).getClass() == parameter.getType()
-        );
+      T t = RethinkDB.getObjectMapper().convertValue(map,pojoClass);
+        return t;
     }
```
Comment (srh) [inline drivers/java/src/main/java/com/rethinkdb/net/Util.java]: The indentation is F'd up here.

**Comment by srh** [comment]
Can somebody explain what the benefit of this patch is?

**Comment by PeterKnego** [comment]
@srh Current Java driver has a limited ability do map values provided by DB (basically a JSON)  to Java objects (you don't wan to marshall/demarshall data coming/going to DB by hand, right?). 

This patch would provide a much better mapping abilities, as provided by Jackson, IMO the best JSON-to-Java-object mapper there is. So this is a good patch that would provide a much better functionality.

However, this will change current API (the data returned/sent to driver will be different) and will break things for current users. As a basic rule minor updates should not change API contract, even major updates should be backwards compatible.

My suggestion would be to make this optional via a config setting and announce this change, than later change make it default when major number is bumped.

**Comment by danielcompton** [comment]
> However, this will change current API (the data returned/sent to driver will be different) and will break things for current users.

My understanding was that the Java driver was still kind of in beta. It's also relatively new, and missing features like async responses which means that many people aren't able to adopt it. I suspect that the number of people who would be affected *and* wouldn't have the engineering resources to update this would be low.

> As a basic rule minor updates should not change API contract, even major updates should be backwards compatible.

That's a good general principle, but may not work with the RethinkDB drivers, as they are versioned against the RethinkDB version, rather than using semver on their own. The RethinkDB protocol is also very stable and RethinkDB is backwards compatible with older protocols, so there is nothing stopping people from relying on the older driver while they migrate.

Especially considering the reduced manpower available for RethinkDB, anything that can reduce maintenance burden is a big win in my book.

**Comment by mcjp78** [comment]
I am conducting tests and I have the requested changes ready to commit.
@PeterKnego, upon looking through the serialization, the JSON actually won't change.

**Comment by srh** [comment]
Please see https://github.com/srh/rethinkdb/tree/sam/javajackson which has a style fix commit.

**Comment by mcjp78** [comment]
@srh Sorry, I didn't see any official style guid :P. Personally, I use allman but I didn't know what here. anyway, can you explain to me why this matters? I am a bit of anstatte

**Comment by chrisvariety** [comment]
@mcjp78 When you add code to an existing codebase you will want to match the existing conventions so all your code looks and reads the same as all the code around it. Doing so will make the codebase one cohesive unit. You'll find this in every open source project, and the majority of closed source projects as well.

I believe RethinkDB has an official style guide though I'm not sure it is public at the moment. In lieu of an official style guide, it should be easy to match the other code already existing in the repository.

**Comment by srh** [comment]
I thought that Javadoc comments don't mean anything (to the Javadoc tool) inside a function body.  At least a quick google suggested that.

I'm not aware of a Java-specific RethinkDB style guide, but having the statements in a block be at the same level of indentation is a more or less universal standard for the same reason that they're indented in the first place.  Regarding spacing, one of the following happens:  Somebody goes and cleans it up later, polluting git blame history.  Or, before that, somebody reading the code will see the weird style, think to themselves, "Oh what is up with this, I better check this out," and start scrutinizing it for quality problems.  So now they're doing work that doesn't need to be done, because of a false bad code smell.  Good code should smell good, so that bad-smelling code is a more reliable signal.

**Comment by mcjp78** [comment]
Ah I see, I will fix the styling then

**Comment by mcjp78** [comment]
@srh  Would you mind checking again to make sure that I fixed everything?

**Comment by marshall007** [comment]
LGTM - @srh or @deontologician can you guys glance over this once more?

**Comment by deontologician** [comment]
LGTM

**Comment by srh** [comment]
Looks good to me but my opinion on Java code isn't worth much.

**Comment by deontologician** [comment]
Ok merging

**Pull request merged**